lib/publiccloud/basetest.pm

```diff
@@ -107,39 +107,61 @@ sub finalize {
+    diag($start_text . '$self->{result}=' . $self->{result}) if ($self->{result});
 
     if ($self->{run_args}) {
-        if ($self->{run_args}->{my_instance}) {
-            my $dumpable_instance = Storable::dclone($self->{run_args}->{my_instance});
-            $dumpable_instance->{provider}->{provider_client}->{credentials_file_content} = '******';
-            diag('Public Cloud finalize: $self->{run_args}->{my_instance}=' . Dumper($dumpable_instance));
-        }
+        diag($start_text . 'checks starting in run_args structure (ref.:$self->{run_args}):');
+        # check Provider
         if ($self->{run_args}->{my_provider}) {
             my $dumpable_provider = Storable::dclone($self->{run_args}->{my_provider});
             $dumpable_provider->{provider_client}->{credentials_file_content} = '******';
-            diag('Public Cloud finalize: $self->{run_args}->{my_provider}=' . Dumper($dumpable_provider));
+            diag($start_text . '$self->{run_args}->{my_provider}= ' . Dumper($dumpable_provider));
+        } else {
+            my $local_text = $start_text . 'Provider is undefined or incomplete (ref.: $self->{run_args}->{my_provider})';
+            record_info('UNDEF. provider', $local_text);
+            diag($local_text);
         }
+        # check Instance
+        if ($self->{run_args}->{my_instance}) {
+            my $dumpable_instance = Storable::dclone($self->{run_args}->{my_instance});
+            $dumpable_instance->{provider}->{provider_client}->{credentials_file_content} = '******';
+            diag($start_text . '$self->{run_args}->{my_instance}=' . Dumper($dumpable_instance));
+            $upload = 1;
+        } else {
+            my $local_text = $start_text . 'Instance is undefined or incomplete (ref.: $self->{run_args}->{my_instance})';
+            record_info('UNDEF. instance', $local_text);
+            diag($local_text);
+        }
+    } else {
+        $start_text .= 'running arguments undefined (ref.: $self->{run_args}).';
+        record_info('FAILED finalize', $start_text);
+        diag($start_text . 'Probably early errors/faults.');
     }
-
     # currently we have two cases when teardown of instance will be skipped:
     # 1. Job should have 'PUBLIC_CLOUD_NO_TEARDOWN' variable
     if (get_var('PUBLIC_CLOUD_NO_TEARDOWN')) {
-        diag('Public Cloud finalize: The test has PUBLIC_CLOUD_NO_TEARDOWN variable.');
-        eval { $self->_upload_logs() } or record_info('FAILED', "\$self->_upload_logs() failed -- $@", result => 'fail');
+        my $ret;
+        diag($start_text . 'The test has PUBLIC_CLOUD_NO_TEARDOWN variable.');
         upload_asset(script_output('ls ~/.ssh/id* | grep -v pub | head -n1'));
-        return;
+        if ($upload) {
+            $ret = $self->_upload_logs();
+            record_info('FAILED upload_logs', $start_text . "Failed _upload_logs no teardown (ref.: \$self->_upload_logs) --\n $@")
```

Comment:
why FAILED ? it is unconditional so even if upload_logs was successful you will still write FAILED box this is confusing